sprocs/questions_select_templates.sql

```diff
@@ -0,0 +1,27 @@
+DROP FUNCTION IF EXISTS questions_select_templates(bigint);
+
+-- Returns the IDs of template questions (if any) for the given question_id
+-- Recursively finds all templates in the chain
+
+CREATE OR REPLACE FUNCTION
+    questions_select_templates (
+        IN question_id bigint
+    ) RETURNS TABLE (id bigint)
+AS $$
+WITH RECURSIVE template_questions AS (
+    -- non-recursive term that finds the ID of the template question (if any) for question_id
+    SELECT tq.*
+    FROM
+        questions AS q
+        JOIN questions AS tq ON (tq.qid = q.template_directory AND tq.course_id = q.course_id)
+    WHERE q.id = question_id
+    -- required UNION for a recursive WITH statement
+    UNION
+    -- recursive term that references template_questions again
+    SELECT tq.*
+    FROM
+        template_questions AS q
```

Comment:
Whoa - am I reading correctly that a recursive function is implemented as a sort of pseudo-table? Never seen a recursive PG function before!

Reply:
Yep! They are pretty neat. They are perfect for self-referential data structures like graphs (or the `question->template_question->template_question->...` pointer chasing). Traditionally graphs are a real pain to store in SQL, but with these recursive queries they become straightforward.